app.py

```diff
@@ -649,17 +649,32 @@ def narrative_services() -> List[dict]:
     narr_services = []
     prespawn_pre = cfg['container_name_prespawn'].format('')
     narr_pre = cfg['container_name'].format('')
+    try:
+        narr_activity = get_narr_activity_from_db()
+    except Exception as e:
+        logger.critical({"message": "Could not get data from database for narrative_status, faking last_seen: {}".format(repr(e))})
+        narr_activity = None
+    
+    try:
+        suffix = stack_suffix()
+    except:
+        suffix = ""
+    
     for name in narr_names:
         if name.startswith(prespawn_pre):
             info = {"state": "queued", "session_id": "*", "instance": name, 'last_seen': time.asctime() }
         else:
             user = name.replace(narr_pre, "", 1)
             info = {"instance": name, "state": "active", "session_id": user}
-            try:
-                info['last_seen'] = time.asctime()
-            except Exception as ex:
-                logger.critical({"message": "Error: adding last_seen field to status message", "error": repr(ex),
-                                 "container": name})
+            if narr_activity:
+                try:
+                    info['last_seen'] = time.asctime(time.gmtime(narr_activity[name+suffix]))
```

Comment:
Does this work?  IIRC, the keys in narr_activity are like 'narrative-userid-hash', with no Rancher environment suffix.

Reply:
I thought that at first too, but then when I dumped out the keys from the DB load, they included the suffix. So I had to add some code to append the suffix before it would work.
```
{"name": "root", "level": "DEBUG", "message": "Activity after updated from traefik: ", "type": "gunicorn_log", "timestamp": "2020-06-26T06:05:01.529Z", "narr_activity": "{'narrative-pranjan77-05fd1a_traefik': '2020-06-26T06:05:01.287332', 'narrative-slebras-130ca3_traefik': '2020-06-26T06:05:01.523123', 'narrative-dylan-dd378c_traefik': '2020-06-26T06:05:01.233251', 'narrative-psdehal-b0ef0d_traefik': '2020-06-26T06:05:01.408329'}", "container": "8152f633deda"}
{"name": "root", "level": "DEBUG", "message": "Saving new narr_activity to database: [('narrative-pranjan77-05fd1a_traefik', 1593151501.2873318), ('narrative-slebras-130ca3_traefik', 1593151501.5231233), ('narrative-dylan-dd378c_traefik', 1593151501.2332513), ('narrative-psdehal-b0ef0d_traefik', 1593151501.4083295)]", "type": "gunicorn_log", "timestamp": "2020-06-26T06:05:01.529Z", "container": "8152f633deda"}

```

Reply:
I just looked further, and you are correct.  The issue I was seeing is that /reaper/ hasn't been called before /narrative_status/ so the session doesn't yet exist in the sqlite db.